### PR TITLE
feat(staticlark): Dataflow analysis for sensitive data

### PR DESCRIPTION
**Pull request by dustmop**
An early implementation of dataflow analysis for starlark. Some caveats:
* control flow not being used, only linear flow is analyzed
* no lexical scoping nor inner functions
* axioms need to be populated

### CODE REVIEW

File: transform/staticlark/analyze.go [hunk 1]
```diff
@@ -24,10 +24,18 @@ func AnalyzeFile(filename string) ([]Diagnostic, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
 	// Return any unused functions
 	// TODO(dustmop): As more analysis steps are introduced, refactor this
 	// into a generic interface that creates Diagnostics
-	return callGraph.findUnusedFuncs(), nil
+	unusedDiags := callGraph.findUnusedFuncs()
+	return append(dataflowDiags, unusedDiags...), nil
```
Comment (b5) [inline transform/staticlark/analyze.go:38]: delighted to see the `Diagnostic` struct already paying dividends here

File: transform/staticlark/environment.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	"strings"
+
+	"go.starlark.net/syntax"
+)
+
+const sensitiveVarName = "$"
+
+// environment is a collection of variable bindings, and the
+// list of other bindings that it derives value from
+// TODO(dustmop): Handle lexical scoping
```
Comment (b5) [inline transform/staticlark/environment.go:15]: what's the plan for scoping? make the environment struct recursive, or a flat struct that applies scopes to string keys?

File: transform/staticlark/dataflow_test.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+			sensitiveReturn: true,
+		},
+		"dangerous": &funcNode{
+			dangerousParams: []bool{true, false},
+			reasonParams: []reason{
+				reason{lines: []string{"assume it is dangerous"}},
+				reason{},
+			},
+		},
+	}
```
Comment (b5) [inline transform/staticlark/dataflow_test.go:30]: This test was super helpful for getting my head around axioms

**Review by b5 — APPROVED** [review]
This is great. I don't want to interrupt with design critique b/c this feels like a midpoint more than a refactoring moment. I find your selected terminology great, really helps lever into the complexity of the topic

File: transform/staticlark/environment.go (hunk 2 above)
Comment (dustmop) [reply transform/staticlark/environment.go:15]: Yes, add a stack to the environment, pushing and pulling as scopes are entered and left. Each binding needs a unique slot, and the name corresponds to that slot based upon which scope it is looked up in. Same type of structure used in a compiler or interpreter.

**Pull request merged**